client-library-otel/src/instrumentation.ts

```diff
@@ -0,0 +1,165 @@
+  },
+};
+
+// // Type hack that makes our middleware types play nicely with Hono types
+// type RouterRoute = {
+//   method: string;
+//   path: string;
+//   // We can't use the type of a handler that's exported by Hono for some reason.
+//   // When we do that, our types end up mismatching with the user's app!
+//   //
+//   // biome-ignore lint/complexity/noBannedTypes:
+//   handler: Function;
+// };
+
+export function instrument(app: Hono, config?: FpxConfigOptions) {
+  return new Proxy(app, {
+    get(target, prop, receiver) {
+      const value = Reflect.get(target, prop, receiver);
+      if (prop === "fetch" && typeof value === "function") {
+        const originalFetch = value as Hono["fetch"];
+        return async function fetch(
+          request: Request,
+          env: unknown,
+          executionContext: ExecutionContext | undefined,
+        ) {
+          // NOTE - We used to have a handy default for the fpx endpoint, but we need to remove that,
+          //        so that people won't accidentally deploy to production with our middleware and
+          //        start sending data to the default url.
+          const endpoint = (env as Record<string, string | null>).FPX_ENDPOINT;
+          const isEnabled = !!endpoint;
+
+          if (!isEnabled) {
+            return await originalFetch(request, env, executionContext);
+          }
+
+          const serviceName =
+            (env as Record<string, string | null>).FPX_SERVICE_NAME ??
+            "unknown";
+
+          // Patch the related functions to monitor
+          const {
+            monitor: { fetch: monitorFetch, logging: monitorLogging },
+          } = mergeConfigs(defaultConfig, config);
+          monitorLogging && patchConsole();
+          monitorFetch && patchFetch();
+
+          const provider = setupTracerProvider({
+            serviceName,
+            endpoint,
+          });
+
+          // Enable tracing for waitUntil
+          const patched = executionContext && patchWaitUntil(executionContext);
+          const promises = patched?.promises ?? [];
+          const proxyExecutionCtx = patched?.proxyContext ?? executionContext;
+
+          const measuredFetch = measure(
+            {
+              name: "route",
+              onStart: (span, [request]) => {
+                span.setAttributes(getRequestAttributes(request));
+              },
+              onEnd: async (span, response) => {
+                const attributes = await getResponseAttributes(
+                  (await response).clone(),
+                );
+                span.setAttributes(attributes);
+              },
+            },
+            originalFetch,
+          );
+
+          const result = await measuredFetch(request, env, proxyExecutionCtx);
+
+          // Make sure all promises are resolved before sending data to the server
+          if (proxyExecutionCtx) {
+            proxyExecutionCtx.waitUntil(
+              Promise.all(promises).finally(() => {
```

Comment:
`Promise.allSettled` instead of `all`? i think `Promise.all` will reject as soon as *any* promise rejects, which means the array of promises will be considered settled and `finally` will execute

`allSettled` makes sure that everything resolves/rejects, and then the `finally` handler executes

Reply:
Oh nice one. I didn't know about that one